### PR TITLE
Mirror pod without OwnerReference should not be created

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR updates the NodeRestriction admission controller to reject mirror pods missing owner reference.

Ref: https://github.com/kubernetes/enhancements/issues/1314

**Which issue(s) this PR fixes**:
Fixes #

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/integration/auth/node_test.go [hunk 1]
```diff
@@ -224,13 +224,26 @@ func TestNodeAuthorizer(t *testing.T) {
 
 	createNode2MirrorPod := func(client clientset.Interface) func() error {
 		return func() error {
-			_, err := client.CoreV1().Pods("ns").Create(context.TODO(), &corev1.Pod{
+			nodeName := "node2"
```
Comment (tallclair) [inline test/integration/auth/node_test.go:227]: nit
```suggestion
			const nodeName = "node2"
```

File: test/integration/auth/node_test.go [hunk 2]
```diff
@@ -492,19 +501,21 @@ func TestNodeAuthorizer(t *testing.T) {
...
+	expectNotFound(t, deleteNode2MirrorPod(node1Client))
+	// create a pod as an admin to add object references
+	expectAllowed(t, createNode2NormalPod(superuserClient))
+
 	expectAllowed(t, createNode2MirrorPod(node2Client))
 	expectAllowed(t, deleteNode2MirrorPod(node2Client))
 	expectAllowed(t, createNode2MirrorPod(node2Client))
 	expectAllowed(t, createNode2MirrorPodEviction(node2Client))
-	expectAllowed(t, createNode2(node2Client))
-	expectAllowed(t, updateNode2Status(node2Client))
```
Comment (tallclair) [inline test/integration/auth/node_test.go:500]: why did you move a bunch of the expectations around?

File: plugin/pkg/admission/noderestriction/admission.go [hunk 3]
```diff
@@ -224,6 +224,9 @@ func (p *Plugin) admitPodCreate(nodeName string, a admission.Attributes) error {
 	if len(pod.OwnerReferences) > 1 {
 		return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with a single owner reference set to itself", nodeName))
 	}
+	if len(pod.OwnerReferences) == 0 {
+		return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with owner reference set to itself", nodeName))
```
Comment (tallclair) [inline plugin/pkg/admission/noderestriction/admission.go:228]: nit
```suggestion
		return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with an owner reference set to itself", nodeName))
```

File: test/integration/auth/node_test.go (hunk 2 above)
Comment (tedyu) [reply test/integration/auth/node_test.go:500]: The existing order is flawed.
e.g. Node2 needs to be created ahead of pod being created on node2.
Otherwise we would get:
```
    TestNodeAuthorizer: wait.go:211: unexpected response, will retry: nodes "node2" not found
    TestNodeAuthorizer: wait.go:211: unexpected response, will retry: nodes "node2" not found
    TestNodeAuthorizer: wait.go:211: unexpected response, will retry: nodes "node2" not found
```

**Comment by tedyu** [comment]
@timstclair 
How about separating the test changes into another PR (and admit it into 1.19) ?

The order of operations in the test should be adjusted to reflect practical usage.

**Comment by tallclair** [comment]
sgtm.

**Comment by tedyu** [comment]
@tallclair 
Looks like 1.19 would be released soon.

When you have time, please give this another look.

**Comment by tedyu** [comment]
@tallclair 
Can you take another look ?

File: plugin/pkg/admission/noderestriction/admission.go [hunk 4]
```diff
@@ -224,6 +224,9 @@ func (p *Plugin) admitPodCreate(nodeName string, a admission.Attributes) error {
 	if len(pod.OwnerReferences) > 1 {
 		return admission.NewForbidden(a, fmt.Errorf("node %q can only create pods with a single owner reference set to itself", nodeName))
 	}
+	if len(pod.OwnerReferences) == 0 {
```
Comment (tallclair) [inline plugin/pkg/admission/noderestriction/admission.go:227]: nit: else if

Comment (tedyu) [reply plugin/pkg/admission/noderestriction/admission.go:227]: The previous if block has return statement. So is this if block.

It seems else is not needed.

**Comment by tedyu** [comment]
@neolit123
Can you approve this ?

Thanks
